### DISCUSSION TITLE
Is it possible to enable/disable menu items upon command execution

### DISCUSSION THREAD

**Discussion by sboudreaux96** [eclipse-theia/theia#9562] [Q&A]
Hi - I am trying to create an extension that contains a command to specifically enable/disable menu items in Theia. Is this possible? I would also like to know if the same can be done as a toggleable feature (like a checkbox)?

**Comment by vince-fugnitto** [comment]
@sboudreaux96 you can use the [MenuModelRegistry](https://eclipse-theia.github.io/theia/docs/next/classes/core.menumodelregistry-1.html) to register and un-register menus (which can be called by the checkbox or executing a command).

For example:

https://github.com/vince-fugnitto/theia-custom-menu/blob/4e1c993cabbff1d7061024ecb3ce8872ba63ebb6/theia-custom-menu/src/browser/theia-custom-menu-contribution.ts#L29-L39

![gs-toggle](https://user-images.githubusercontent.com/40359487/121079176-d3188f00-c7a7-11eb-8f58-c0be0a35b68a.gif)

There are limitations however, for example in the `electron` target I do not believe that such dynamic menus are supported:
- https://github.com/eclipse-theia/theia/issues/446

**Reply by sboudreaux96** [reply]
@vince-fugnitto I have tried implementing my own version of this but keep getting exit code 1 due to the below error:
src/browser/TheiaCustomMenu-frontend-module.ts(4,46): error TS2305: Module '"./TheiaCustomMenu-contribution"' has no exported member 'TheiaCustomMenuMenuContribution'.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.

    at C:\Users\SBoudreaux\TheiaAgain\theia\CustomWidgets\node_modules\execa\index.js:236:11
    at processTicksAndRejections (internal/process/task_queues.js:97:5)  
(node:44400) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 4)
(node:44400) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.

**Reply by vince-fugnitto** [reply]
@sboudreaux96 it is difficult to tell without seeing the actual code but it looks like you forgot to export `TheiaCustomMenuMenuContribution` when using it in other files like the `*-module.ts` (to perform binding).

**Reply by sboudreaux96** [reply]
@vince-fugnitto I believe I have bound the contributions correctly below (frontend-module.ts):

```ts
import { TheiaCustomMenuCommandContribution } from './TheiaCustomMenu-contribution';
import {
    CommandContribution,
    MenuContribution
} from "@theia/core/lib/common";
import { ContainerModule } from "inversify";

export default new ContainerModule(bind => {
    bind(CommandContribution).to(TheiaCustomMenuCommandContribution);
    bind(MenuContribution).to(TheiaCustomMenuCommandContribution);
});
```

Below is my version of the contribution.ts file:

```ts
import { injectable, inject } from "inversify";
import { Command, CommandContribution, CommandRegistry, MenuContribution, MenuModelRegistry, MessageService } from "@theia/core/lib/common";
import { CommonMenus } from "@theia/core/lib/browser";
import { GettingStartedCommand } from '@theia/getting-started/lib/browser/getting-started-contribution'

export const TheiaCustomMenuCommand = {
    id: 'TheiaCustomMenu.command',
    label: "Say Hello"
};

@injectable()
export class TheiaCustomMenuCommandContribution implements CommandContribution, MenuContribution {

    constructor(
        // @inject(MessageService) private readonly messageService: MessageService,
    ) { }

    @inject(MenuModelRegistry)
    protected isEnabled: boolean = true;
    protected _command: Command = GettingStartedCommand;
    menuRegistry: MenuModelRegistry;


    registerCommands(registry: CommandRegistry): void {
        registry.registerCommand(TheiaCustomMenuCommand, {
            execute: () => {
                this.toggleMenu(this.isEnabled);
                this.isEnabled = !this.isEnabled;
            }
        });
    }

    protected toggleMenu(isEnabled: boolean): void {
        if (isEnabled) {
            this.menuRegistry.unregisterMenuAction(this._command.id);
        } else {
            this.menuRegistry.registerMenuAction(CommonMenus.HELP, {
                commandId: this._command.id,
                label: this._command.label,
                order: 'a10'
            });
        }
    }
    registerMenus(menus: MenuModelRegistry): void {
        menus.registerMenuAction(CommonMenus.EDIT_FIND, {
            commandId: TheiaCustomMenuCommand.id,
            label: TheiaCustomMenuCommand.label
        });
    }
}
```

**Reply by vince-fugnitto** [reply]
For one your injection of `MenuModelRegistry` looks incorrect:

```ts
    @inject(MenuModelRegistry)
    protected isEnabled: boolean = true;
    protected _command: Command = GettingStartedCommand;
    menuRegistry: MenuModelRegistry;
 ```

it should be:

```ts
    @inject(MenuModelRegistry)
    menuRegistry: MenuModelRegistry;

    protected isEnabled: boolean = true;
    protected _command: Command = GettingStartedCommand;
````

**Reply by sboudreaux96** [reply]
I have fixed that, but still giving the same error

**Reply by vince-fugnitto** [reply]
> I have fixed that, but still giving the same error

Can you push your changes to a public repository where I can verify it?
I already have the complete example you can follow: https://github.com/vince-fugnitto/theia-custom-menu

**Reply by sboudreaux96** [reply]
Sure, here is the repo: https://github.com/sboudreaux96/theia-test

**Reply by vince-fugnitto** [reply]
> Sure, here is the repo: https://github.com/sboudreaux96/theia-test

Please note that it is not recommended to fork the main repository and add extensions.
Instead you should really follow the guide from [composing an application](https://theia-ide.org/docs/composing_applications/).

As for the errors, you are missing a dependency to `@theia/getting-started`, and usage of `MessageService` should be removed if it is unused:

```json
"dependencies": {
    "@theia/core": "latest",
    "@theia/getting-started": "latest"
},
 ```

**Reply by sboudreaux96** [reply]
ok that makes sense. I appreciate the guidance and pointing me in the right direction for these! this works!

**Reply by sboudreaux96** [reply]
@vince-fugnitto I have one more question for you: how did you know what dependencies to import? I am currently trying to get to workspace-commands.ts but it seems I cant find out how to import it properly.

**Reply by vince-fugnitto** [reply]
> @vince-fugnitto I have one more question for you: how did you know what dependencies to import? I am currently trying to get to workspace-commands.ts but it seems I cant find out how to import it properly.

@sboudreaux96 search the sources for the commands you require (either by label) and the corresponding `package` it resides in will be the dependency you require. The `workspace-commands` are likely from `@theia/workspace`:

https://github.com/eclipse-theia/theia/blob/f9c031120708e0b4c7ef4afe86210144062ebde8/packages/workspace/src/browser/workspace-commands.ts#L43

There is also the `api` docs: https://eclipse-theia.github.io/theia/docs/next/modules/workspace.workspacecommands-1.html.

**Reply by sboudreaux96** [reply]
@vince-fugnitto that makes sense as well, but when importing the dependency how do you know what to put after "@theia/"?

**Reply by vince-fugnitto** [reply]
> @vince-fugnitto that makes sense as well, but when importing the dependency how do you know what to put after "@theia/"?

I'm not sure what is confusing, the repository is broken down into multiple extensions under `packages`.
The `packages/workspace` folder contains the `@theia/workspace` extension:

https://github.com/eclipse-theia/theia/blob/f9c031120708e0b4c7ef4afe86210144062ebde8/packages/workspace/package.json#L2

Which is published to `npm`: https://www.npmjs.com/package/@theia/workspace.

**Reply by sboudreaux96** [reply]
@vince-fugnitto so I have tried to import the workspace commands in the same way GettingStartedCommand was and I ended up getting the below error. I have tried rebuilding and updating (and deleting) the corresponding dependency but nothing changes it.

Module not found: Error: Can't resolve '@theia/workspace/src/browser/workspace-commands' in 'C:\Users\SBoudreaux\TheiaAgain\theia\CustomWidgets\TheiaCustomMenu\lib\browser'
resolve '@theia/workspace/src/browser/workspace-commands' in 'C:\Users\SBoudreaux\TheiaAgain\theia\CustomWidgets\TheiaCustomMenu\lib\browser' 
  Parsed request is a module
  using description file: C:\Users\SBoudreaux\TheiaAgain\theia\CustomWidgets\TheiaCustomMenu\package.json (relative path: ./lib/browser)      
    Field 'browser' doesn't contain a valid alias configuration
    resolve as module
      C:\Users\SBoudreaux\TheiaAgain\theia\CustomWidgets\TheiaCustomMenu\lib\browser\node_modules doesn't exist or is not a directory
      C:\Users\SBoudreaux\TheiaAgain\theia\CustomWidgets\TheiaCustomMenu\lib\node_modules doesn't exist or is not a directory
      looking for modules in C:\Users\SBoudreaux\TheiaAgain\theia\CustomWidgets\TheiaCustomMenu\node_modules
        C:\Users\SBoudreaux\TheiaAgain\theia\CustomWidgets\TheiaCustomMenu\node_modules\@theia\workspace doesn't exist
      looking for modules in C:\Users\SBoudreaux\TheiaAgain\theia\CustomWidgets\node_modules
        existing directory C:\Users\SBoudreaux\TheiaAgain\theia\CustomWidgets\node_modules\@theia\workspace
          using description file: C:\Users\SBoudreaux\TheiaAgain\theia\CustomWidgets\node_modules\@theia\workspace\package.json (relative path: .)
            using description file: C:\Users\SBoudreaux\TheiaAgain\theia\CustomWidgets\node_modules\@theia\workspace\package.json (relative path: ./src/browser/workspace-commands)
              no extension
                Field 'browser' doesn't contain a valid alias configuration
                C:\Users\SBoudreaux\TheiaAgain\theia\CustomWidgets\node_modules\@theia\workspace\src\browser\workspace-commands doesn't exist 
              .js
                Field 'browser' doesn't contain a valid alias configuration
                C:\Users\SBoudreaux\TheiaAgain\theia\CustomWidgets\node_modules\@theia\workspace\src\browser\workspace-commands.js doesn't exist
              as directory
                C:\Users\SBoudreaux\TheiaAgain\theia\CustomWidgets\node_modules\@theia\workspace\src\browser\workspace-commands doesn't exist 
      looking for modules in C:\Users\SBoudreaux\TheiaAgain\theia\node_modules
        existing directory C:\Users\Boudreaux\TheiaAgain\theia\node_modules\@theia\workspace
          using description file: C:\Users\SBoudreaux\TheiaAgain\theia\node_modules\@theia\workspace\package.json (relative path: .)
            using description file: C:\Users\SBoudreaux\TheiaAgain\theia\node_modules\@theia\workspace\package.json (relative path: ./src/browser/workspace-commands)
              no extension
                Field 'browser' doesn't contain a valid alias configuration
                C:\Users\SBoudreaux\TheiaAgain\theia\node_modules\@theia\workspace\src\browser\workspace-commands doesn't exist
              .js
                Field 'browser' doesn't contain a valid alias configuration
                C:\Users\SBoudreaux\TheiaAgain\theia\node_modules\@theia\workspace\src\browser\workspace-commands.js doesn't exist
              as directory
                C:\Users\SBoudreaux\TheiaAgain\theia\node_modules\@theia\workspace\src\browser\workspace-commands doesn't exist
      C:\Users\SBoudreaux\TheiaAgain\node_modules doesn't exist or is not a directory
      looking for modules in C:\Users\SBoudreaux\node_modules        
        C:\Users\SBoudreaux\node_modules\@theia\workspace doesn't exist
      C:\Users\node_modules doesn't exist or is not a directory
      C:\node_modules doesn't exist or is not a directory
 @ ../TheiaCustomMenu/lib/browser/TheiaCustomMenu-frontend-module.js 6:37-78
 @ ./src-gen/frontend/index.js 59:31-100

webpack 5.38.1 compiled with 1 error and 44 warnings in 11966 ms


    at C:\Users\SBoudreaux\TheiaAgain\theia\CustomWidgets\node_modules\execa\index.js:236:11
    at processTicksAndRejections (internal/process/task_queues.js:97:5)  
(node:39428) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 6)
(node:39428) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.

**Reply by vince-fugnitto** [reply]
@sboudreaux96 unfortunately I cannot help you each time there is an implementation error with your custom extension. 
Please debug as to what might cause the error, and see if you are importing it correctly (like the `getting-started` commands using the `lib`). As always it is much easier to help if there is a branch I can refer to instead of just the error log.
